Main.hs

```diff
@@ -363,29 +315,69 @@ renderBlocks = unlines . map unlines . sortRanges . filter (not . null . snd)
 defBlock :: CompiledDef -> [Block]
 defBlock def = [ (r, map pp ds) | (r, ds) <- def ]
 
-codePragmas :: [(Range, Code)] -> [Block]
+codePragmas :: [Ranged Code] -> [Block]
 codePragmas code = [ (r, map pp ps) | (r, (Hs.Module _ _ ps _ _, _)) <- code ]
 
-codeBlocks :: [(Range, Code)] -> [Block]
+codeBlocks :: [Ranged Code] -> [Block]
 codeBlocks code = [(r, [uncurry Hs.exactPrint $ moveToTop $ noPragmas mcs]) | (r, mcs) <- code, nonempty mcs]
   where noPragmas (Hs.Module l h _ is ds, cs) = (Hs.Module l h [] is ds, cs)
         noPragmas m                           = m
         nonempty (Hs.Module _ _ _ is ds, cs) = not $ null is && null ds && null cs
         nonempty _                           = True
 
--- exactPrint really looks at the line numbers (and we're using the locations from the agda source
--- to report Haskell parse errors at the right location), so shift everything to start at line 1.
-moveToTop :: Hs.Annotated ast => (ast Hs.SrcSpanInfo, [Hs.Comment]) -> (ast Hs.SrcSpanInfo, [Hs.Comment])
-moveToTop (x, cs) = (subtractLine l <$> x, [ Hs.Comment b (sub l r) str | Hs.Comment b r str <- cs ])
-  where l = Hs.startLine (Hs.ann x) - 1
-        subtractLine :: Int -> Hs.SrcSpanInfo -> Hs.SrcSpanInfo
-        subtractLine l (Hs.SrcSpanInfo s ss) = Hs.SrcSpanInfo (sub l s) (map (sub l) ss)
+-- Checking imports -------------------------------------------------------
 
-        sub :: Int -> Hs.SrcSpan -> Hs.SrcSpan
-        sub l (Hs.SrcSpan f l0 c0 l1 c1) = Hs.SrcSpan f (l0 - l) c0 (l1 - l) c1
+imports :: [Ranged Code] -> [Hs.ImportDecl Hs.SrcSpanInfo]
+imports modules = concat [imps | (_, (Hs.Module _ _ _ imps _, _)) <- modules]
 
-pp :: Hs.Pretty a => a -> String
-pp = Hs.prettyPrintWithMode Hs.defaultMode{ Hs.spacing = False }
+addImports :: [Hs.ImportDecl Hs.SrcSpanInfo] -> [CompiledDef] -> TCM [Hs.ImportDecl ()]
+addImports is defs = do
+  return [importWord64 | not (null defs || any isImportWord64 is)] 
```

Comment:
This will add `import Data.Word (Word64)` to every module with a non-empty set of (Haskell) declarations, but we probably want to do that only when there is relevant code in that module. This seems quite hard to check though (you need a hundred LOC to traverse all possible `Hs.`constructors and check types referred to there), but maybe a SYB-generic function can do that easily enough.

Reply:
Another option is to remember when you generate the `Word64` type.

I think if you can't detect when to import it, it's better to not do anything than to always import. It's not that hard to write the Haskell import yourself and it would be really annoying to have an unnecessary `Data.Word` import everywhere.

Reply:
OK, let's try to do this on the Agda side (i.e. remembering when we generate `Word64`) and if it's not so easy, abandon automatic importing altogether. Any quick suggestion on how to manage custom state in `TCM`, do I really need to wrap it in `WriterT` or is there a convenient way to do this from TCM itself?

Reply:
Yeah, I think `WriterT` is your best bet.

Reply:
Phew, there was a trivial way to do this using SYB's [listify](https://hackage.haskell.org/package/syb-0.7.1/docs/Data-Generics-Schemes.html#v:listify).